client/core: hide order form and display registration status message

fernandoabolafio opened the pull request:
This PR modifies the code so the order form is hidden while the registration payment has not been confirmed. Instead, a message is shown to communicate the status of the confirmation. See the previews for more info.

The full list of changes:

**Core changes:**
- Adds the dex URL to all `feepayment` notifications. This is particularly useful in this case to determine to either show or not the confirmation message in the current market.
- The registration update notification uses now the `regupdated` subject.
- Successful registration note uses now the `regcompleted` subject. So it server to multiple purposes in the frontend side. See 'Site changes' for more info. 
- Refresh the user after completed registration.

**Site changes:**
- ~Added an override handler in the app.js which enables the overriding of notifications subject. This is useful to have a notification serving multiple purposes, such as the 'regcompleted' in this case where it is both displayed in the notifications dropdown and it is used to display the order form once the registration payment has been successful.~
- The ~market~ app page now handles `feepayment` notifications and update the exchanges confirmations as well the `feePending` value.
- Hide 'order form' and display the 'registration status' while the fee payment has not been confirmed. Once it is confirmed, a success message is shown for 5 seconds and the order form is set as visible again.


**UI Preview:**

![Screenshot 2020-05-08 at 12 34 18](https://user-images.githubusercontent.com/14864439/81497868-a56dc780-92c1-11ea-95cb-7952d3607b34.png)
![Screenshot 2020-05-08 at 12 33 36](https://user-images.githubusercontent.com/14864439/81497870-a69ef480-92c1-11ea-809d-a5adb06083f8.png)
![Screenshot 2020-05-08 at 12 33 09](https://user-images.githubusercontent.com/14864439/81497874-a999e500-92c1-11ea-976e-2c33890c5b5f.png)


closes #337

chappjc commented:
@fernandoabolafio Minor conflict to resolve please.

fernandoabolafio commented:
@chappjc thanks for the heads up I am still working on some changes. I will fix all conflicts once I am done.

fernandoabolafio commented on client/webserver/site/src/js/markets.js:
```diff
@@ -225,6 +231,95 @@ export default class MarketsPage extends BasePage {
+    const page = this.page
+
+    page.regStatusDex.textContent = this.market.dex.url
+    function updateHandler (titleContent, messageContent, titleClass) {
```
Using the method approach.

fernandoabolafio commented:
@buck54321 I addressed all of your comments. The code looks much simpler now. Major changes are:
- Instead of tracking the registration status with a new object, I am just using the `user.exchanges`.  The tracking happens only in the App level and the loaded page is notified using a new method called `registrationStatusUpdated`.
- I added `confsRequired` to the exchange object so now it doesn't need to be provided with notifications and the app is aware of this information from the start.

buck54321 commented on client/webserver/site/src/js/basepage.js:
```diff
@@ -5,6 +5,11 @@ export default class BasePage {
     this.notifiers[note.type](note)
   }
 
+  registrationStatusUpdated (dexUrl) {
+    if (!this.onRegistrationStatusUpdated) return
+    this.onRegistrationStatusUpdated(dexUrl)
+  }
```
This is only used on the `MarketsPage`, so it might as well trigger with the `feepayment` notification. Order of operations is

1. `Application` receives `feepayment` notification and sets appropriate fields in the `app.user.exchanges[url]`. 
2. `MarketsPage` receives `feepayment` notification and `MarketsPage.onRegistrationStatusUpdated(note.dexurl)`

fernandoabolafio replied:
I like that! I made the changes as you suggested :)

buck54321 commented on client/webserver/site/src/js/app.js:
```diff
@@ -222,6 +223,30 @@ export default class Application {
+  updateExchangeRegistration (dexUrl, isPaid, confs) {
+    const dex = this.exchanges[dexUrl]
+    dex.feePending = !isPaid
+    dex.confs = confs
```
I'd like to see a change to the `Exchange` struct so that this info isn't lost on refresh. In particular, the `FeePending` field should go away, and the confirmation count added as a pointer field. 

```
type Exchange struct {
  ...
  ConfsRequired uint32  `json:"confsReq"`
  RegConfirms   *uint32 `json:"confs"`  
}
```

Where a `nil` `RegConfirms` means the DEX is ready to go.

My recommended solution is to catch the `FeePaymentNote` in `(*Core).notify` (this'll be the first time we use that pattern, but I've been considering it as possibility for other things too), and update the `RegConfirms` field there, but if you see a better way, that's fine too.

You'll will still want to keep these updates to the browser state as is, but you will have the right number of confirmations after a refresh if you do it that way.

buck54321 replied:
Note updated recommendation.

fernandoabolafio replied:
Great catch on persisting the number of confs across sessions! I implemented as you suggested. One thing that you'll probably notice is that I am not calling `(*Core).refreshUser` after increasing the number of confirmations because that field is a pointer, so changing it into the `dexConnection` is enough to get it updated. Is that ok?

buck54321 replied:
See #379, but yes, you will need to call `refreshUser` any time the `dexConnection.regConfirms` field is updated.

fernandoabolafio replied:
Alright

buck54321 commented on client/webserver/site/src/js/markets.js:
```diff
@@ -225,6 +230,73 @@ export default class MarketsPage extends BasePage {
+      return
+    }
+
+    const waitingMessage = [confirmations, confirmationsRequired]
+      .every(v => typeof v === 'number') ? `${confirmations} / ${confirmationsRequired}` : ''
```
`confirmationsRequired` will always be a number now, I think.

fernandoabolafio replied:
ok, I removed this type check.

buck54321 commented on client/webserver/site/src/js/markets.js:
```diff
@@ -225,6 +230,73 @@ export default class MarketsPage extends BasePage {
+    this.setRegistrationStatusView('Waiting for confirmations...', waitingMessage, 'waiting')
+  }
+
+  setRegistrationStatusVisibility () {
```
Docstring here.

buck54321 commented on client/webserver/site/src/js/doc.js:
```diff
@@ -74,6 +74,10 @@ export default class Doc {
     for (const el of els) el.classList.remove('d-hide')
   }
 
+  static isHidden (el) {
```
Docstring here.

buck54321 commented on client/webserver/site/src/js/markets.js:
```diff
@@ -225,6 +230,73 @@ export default class MarketsPage extends BasePage {
+    }
+  }
+
+  /**
```
One too many asterisks.

fernandoabolafio commented on client/core/core.go:
```diff
@@ -391,10 +393,11 @@ func (c *Core) Exchanges() map[string]*Exchange {
-			URL:        uri,
-			Markets:    dc.markets(),
-			Assets:     dc.assets,
-			FeePending: dc.acct.feePending(),
```
@buck54321 with these changes, the method `(*dexAccount).feePending()` is not being used elsewhere. Should we just remove it?

buck54321 replied:
Yes, please.

buck54321 commented on client/webserver/site/src/js/markets.js:
```diff
@@ -621,6 +687,18 @@ export default class MarketsPage extends BasePage {
+    const dexUrl = note.dexurl
+    if (dexUrl !== this.market.dex.url) return
+    // update local dex
+    this.market.dex = app.exchanges[dexUrl]
```
Looking around now, I'm noticing that `this.market.dex` is only ever used for the `dex.url`. Maybe we should just set the field as `this.market.dexAddr`, which won't change, so we don't have to update the `dex` object after `fetchUser` is called.

fernandoabolafio replied:
hmm. I don't think I follow you on this one. We also use `this.market.dex` to get the `confs` and `confsRequired` and that's the major reason we update the dex value. 
See [here](https://github.com/decred/dcrdex/pull/353/files#diff-12ae7f4ee80365e11e28e25b80759a65R270).

buck54321 commented on client/webserver/site/src/html/markets.tmpl:
```diff
@@ -116,6 +116,21 @@
+                <div class="d-flex flex-column justify-content-center align-items-center">
+                  <span id="regStatusTitle" class="title"></span>
+                  <p>In order to trade at <span id="regStatusDex"></span>, the registration fee payment
+                    needs <span id="confReq"></span> confirmations. The number of confirmations received will appear below once it
+                    is available.
+                  </p>
```
~~The number of confirmations received will appear below once it is available.`~~. 

It's always available now, I think.

buck54321 commented on client/rpcserver/handlers.go:
```diff
@@ -589,8 +589,11 @@ Registration is complete after the fee transaction has been confirmed.`,
+		},
+		"confsrequired": (int) The number of confirmations needed for the 
+		registration fee payment
+		"confs" (int/null): The current number of confirmations for the registration
+		fee payment or null when the payment is confirmed
```
Consider using a `json:"confs,omitempty"` here. Maybe say something like `... only present during the registration process.`

buck54321 commented on client/core/notification.go:
```diff
@@ -100,19 +100,19 @@ type Notification interface {
-	ConfirmationsRequired uint32 `json:"confirmationsrequired,omitempty"`
-	Confirmations         uint32 `json:"confirmations,omitempty"`
+	Confirmations uint32 `json:"confirmations,omitempty"`
+	DexURL        string `json:"dexurl,omitempty"`
```
See #374, but use just `DEX` and `json:"dex"` here for now. DEX field should never be empty. `omitempty` is also confusing for `Confirmations`, since `0` is a valid value. Maybe a `*uint32` instead?

fernandoabolafio replied:
Makes sense :)

buck54321 commented on client/core/notification.go:
```diff
@@ -100,19 +100,19 @@ type Notification interface {
 }
 
-func newFeePaymentNote(subject, details string, severity db.Severity) *FeePaymentNote {
+func newFeePaymentNote(subject, details string, severity db.Severity, url string) *FeePaymentNote {
```
~~`url`~~ 🡒 `dexAddr` a couple of places. We're still a little mixed, but I don't wan't to make more work for #374

fernandoabolafio commented:
@buck54321 thanks again for the comments! All the comments were addressed with the exception of this one: https://github.com/decred/dcrdex/pull/353#discussion_r428275304 where I left a question for you.

chappjc commented:
url -> host rename created a couple very minor conflicts.

chappjc commented on client/webserver/site/src/js/markets.js:
```diff
@@ -658,6 +726,18 @@ export default class MarketsPage extends BasePage {
+   */
+  handleFeePayment (note) {
+    const dexAddr = note.dex
+    if (dexAddr !== this.market.dex.url) return
```
Note that `dex.url` is now `dex.host`, with some related fixes in PR #399 .

chappjc commented:
PR #399 fixing some url/host variable name issues now merged.  Should be able to resolve and update to use dex.host instead of dex.url in your new code.  Thanks and sorry for the repeated conflicts!  Lots of concurrent work on the same files right now.

fernandoabolafio commented on client/core/notification.go:
```diff
@@ -100,20 +100,20 @@ type Notification interface {
 	return &FeePaymentNote{
-		Notification: db.NewNotification("fee payment", subject, details, severity),
+		Notification: db.NewNotification("feepayment", subject, details, severity),
+		Dex:          addrHost(dexAddr),
```
@chappjc thanks for the heads up. I updated the code to use the dex host instead. The `feepayment` is also sending the host by making sure to transform `dexAddr` using the `addrHost` helper.

buck54321 commented on client/webserver/site/src/js/app.js:
```diff
@@ -239,6 +240,40 @@ export default class Application {
+  }
+
+  /*
+   * handleFeePaymentNote is the handler for the 'feepayment'-type notificaetion, which
```
notificaetion

fernandoabolafio commented:
@buck54321 I addressed the changes you requested: 
- Added a method called `  resolveOrderFormVisibility` which is used every time another methods wants to update the `orderForm` visibility. The conditions which changes the form visibility are defined inside `resolveOrderFormVisibility`.
- Simplified a bit the code at the end of `setMarket` and merged the 'loaderMsg' functions (hide/show) into a single one.
- Add to 'query methods' `hasFeePending` and `assetsAreSupported` which are used to ease the readability and encapsulate logic we are reusing. 
- Style tweak:
![Screenshot 2020-05-26 at 18 52 39](https://user-images.githubusercontent.com/14864439/82931255-ea168580-9f86-11ea-985f-6e63e0da6b71.png)
![Screenshot 2020-05-26 at 18 53 39](https://user-images.githubusercontent.com/14864439/82931247-e682fe80-9f86-11ea-9947-17660a07280a.png)

chappjc commented on client/core/core.go:
```diff
@@ -485,11 +487,13 @@ func (c *Core) Exchanges() map[string]*Exchange {
+			FeePending:    dc.acct.feePending(),
+			Connected:     dc.connected,
+			ConfsRequired: uint32(dc.cfg.RegFeeConfirms),
+			RegConfirms:   dc.regConfirms,
```
`refreshUser`=>`Exchanges` reading `dc.regConfirms` can happen in many places, but `regConfirms` is not guarded (`connMtx` is for `Core.conns`, not for these fields of `dexConnection`).

chappjc commented on client/core/core.go:
```diff
@@ -963,6 +967,15 @@ func (c *Core) Register(form *RegisterForm) error {
+		return
+	}
+
+	dc.regConfirms = &regConfs
```
Possible concurrent write if refreshUser or Exchanges is called elsewhere.

chappjc commented on client/core/core.go:
```diff
@@ -982,10 +997,12 @@ func (c *Core) verifyRegistrationFee(wallet *xcWallet, dc *dexConnection, coinID
 			} else {
 				details := fmt.Sprintf("You may now trade at %s", dc.acct.url)
-				c.notify(newFeePaymentNote("Account registered", details, db.Success))
+				dc.regConfirms = nil
```
Possible concurrent write.

chappjc commented on client/core/types.go:
```diff
@@ -163,11 +163,13 @@ func (m *Market) marketName() string {
+	FeePending    bool                  `json:"feePending"`
+	Connected     bool                  `json:"connected"`
+	ConfsRequired uint32                `json:"confsrequired"`
+	RegConfirms   *uint32               `json:"confs,omitempty"`
```
Only needs to be a pointer in `Exchange` for purposes of `omitempty`, but it's a bit awkward as a pointer in `dexConnection`.  The alternative would be to have -1 or 0 be special values in `dexConnection.regConfirms` that indicate to write `nil` to `Exchange.RegConfirms` in `(*Core).Exchanges`.

fernandoabolafio commented on client/core/core.go:
```diff
@@ -68,6 +71,9 @@ type dexConnection struct {
 	// connected is a best guess on the ws connection status.
 	connected bool
+
+	regConfMtx  sync.RWMutex
+	regConfirms uint32
```
Added a `sync.RWMutex` to be used when writing/reading `regConfirms`

fernandoabolafio commented on client/core/core.go:
```diff
@@ -115,6 +121,25 @@ func (dc *dexConnection) markets() map[string]*Market {
 	return dc.marketMap
 }
 
+// getRegConfirms returns the number of confirmations received for the
+// dex registration or nil if the registration is completed
+func (dc *dexConnection) getRegConfirms() *uint32 {
+	dc.regConfMtx.RLock()
+	defer dc.regConfMtx.RUnlock()
+	if dc.regConfirms == regConfirmationsPaid {
+		return nil
+	}
+	return &dc.regConfirms
+}
```
The `getRegConfirms` which returns an `uint32` pointer which can be used in the `core.Exchanges`. It will return the number of confirmations or `nil` if the registration has been completed. 
Notice that different from your suggestion I used a high number called `regConfirmationsPaid` to indicate that the confirmation is completed.
I could not use `0` because it just means that the registration is ongoing and no confirmation has been received so far. Using -1  requires converting the `dexConnection.regConfirms` to `int32` which hurts the semantic of the code and would also for me to convert the `Exchange` field to `int` at some point.

fernandoabolafio commented on client/core/core.go:
```diff
@@ -115,6 +121,25 @@ func (dc *dexConnection) markets() map[string]*Market {
+
+// setRegConfirms sets the number of confirmations received
+// for the dex registration
+func (dc *dexConnection) setRegConfirms(confs uint32) {
```
Simple method to update `regConfirms` which deals with the synchronization internally.

fernandoabolafio commented on client/core/core.go:
```diff
@@ -36,6 +36,9 @@ const (
 	regFeeAssetSymbol = "dcr" // Hard-coded to Decred for registration fees, for now.
+	// high number to be assigned to 'regConfirms' in 'dexConnection'
+	// when the registration is completed
+	regConfirmationsPaid uint32 = 9999
```
The 'PNOOMA' number used to represent that the confirmation is completed.

chappjc replied:
OK, although note this is actually realistic for many ccs.  e.g. ETH: https://blog.coinbase.com/announcing-new-confirmation-requirements-4a5504ba8d81

chappjc commented on client/core/core.go:
```diff
@@ -497,7 +520,7 @@ func (c *Core) Exchanges() map[string]*Exchange {
 			Connected:     dc.connected,
 			ConfsRequired: uint32(dc.cfg.RegFeeConfirms),
-			RegConfirms:   dc.regConfirms,
+			RegConfirms:   dc.getRegConfirms(),
```
When code accesses dereferences this `Exchange.RegConfirms` pointer, it's accessing `dexConnection.regConfirms` because`getRegConfirms` does `return &dc.regConfirms`.  So this is basically sidestepping the locking as when another goroutine calls `dc.setRegConfirms` it is not locked anymore.

You need to either have `getRegConfirms` return a plain `uint32` or make a copy inside the function and return a pointer to the copy (`confs := dc.regConfirms; return &confs`).

fernandoabolafio replied:
Makes sense! I changed it to return the copy instead.
